### PR TITLE
gitserver: able to clone and fetch Perforce depots

### PR DESCRIPTION
**Pull request by unknwon**
This PR adds the ability to actually clone and fetch Perforce Depots as Git repositories, include handling of its session-based authentication.

Part of #16703

### TODO

- [x] Redact username and password from repository "Settings > Remote repository URL"
- [x] Handle interactive password input: https://stackoverflow.com/questions/44471749/golang-enter-ssh-sudo-password-on-prompt-or-exit
- [x] Tests

### CODE REVIEW

File: cmd/frontend/graphqlbackend/repository_mirror.go [hunk 1]
```diff
@@ -58,7 +58,7 @@ func (r *repositoryMirrorInfoResolver) repoUpdateSchedulerInfo(ctx context.Conte
 
 // TODO(flying-robot): this regex and the majority of the removeUserInfo function can
 // be extracted to a common location in a subsequent change.
-var nonSCPURLRegex = lazyregexp.New(`^(git\+)?(https?|ssh|rsync|file|git)://`)
+var nonSCPURLRegex = lazyregexp.New(`^(git\+)?(https?|ssh|rsync|file|git|perforce)://`)
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/repository_mirror.go:61]: To redact credentials from repository settings UI.

File: cmd/gitserver/server/vcs_syncer.go [hunk 2]
```diff
@@ -88,20 +95,172 @@ func (s *GitRepoSyncer) FetchCommand(ctx context.Context, url string) (cmd *exec
...
+	password, _ = url.User.Password()
+	return url.User.Username(), password, url.Host, url.Path, nil
+}
+
+// ping sends one message to the Perforce Server to check connectivity.
+func (s *PerforceDepotSyncer) ping(ctx context.Context, host, username string) error {
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	cmd := exec.CommandContext(ctx, "p4", "ping", "-c", "1")
```
Comment (unknwon) [inline cmd/gitserver/server/vcs_syncer.go:127]: By default it sends out 10 messages, which is a waste IMO.

File: cmd/gitserver/server/vcs_syncer.go [hunk 3]
```diff
@@ -88,20 +95,172 @@ func (s *GitRepoSyncer) FetchCommand(ctx context.Context, url string) (cmd *exec
...
+}
+
+// pingWithLogin attempts to ping the Perforce Server and performs a login operation when needed.
+func (s *PerforceDepotSyncer) pingWithLogin(ctx context.Context, host, username, password string) error {
+	// Attempt to check connectivity, may be prompted to login (again)
+	err := s.ping(ctx, host, username)
+	if err == nil {
+		return nil // The ping worked, session still validate for the user
+	} else if !strings.Contains(err.Error(), "Your session has expired, please login again.") &&
+		!strings.Contains(err.Error(), "Perforce password (P4PASSWD) invalid or unset.") {
```
Comment (unknwon) [inline cmd/gitserver/server/vcs_syncer.go:198]: p4 is a CLI after all, so its error message is more user-friendly than machine-friendly. Happy to improve if anyone has a better pattern.

File: schema/perforce.schema.json [hunk 4]
```diff
@@ -25,7 +25,7 @@
     "depots": {
       "description": "Depots can have arbitrary paths, e.g. a path to depot root or a subdirectory.",
       "type": "array",
-      "items": { "type": "string", "pattern": "^\\/[\\/\\w]+/$" },
+      "items": { "type": "string", "pattern": "^\\/[\\/\\S]+\\/$" },
```
Comment (unknwon) [inline schema/perforce.schema.json:28]: Changed because `\w` does not allow hyphens which are valid as a depot name.

File: cmd/gitserver/server/server.go [hunk 5]
```diff
@@ -876,9 +876,12 @@ func (s *Server) cloneRepo(ctx context.Context, repo api.RepoName, opts *cloneOp
 		if err != nil {
 			return errors.Wrap(err, "get clone command")
 		}
+		if cmd.Env == nil {
+			cmd.Env = os.Environ()
```
Comment (ryanslade) [inline cmd/gitserver/server/server.go:880]: Why was this change required?

File: cmd/gitserver/server/vcs_syncer.go [hunk 6]
```diff
@@ -88,20 +95,172 @@ func (s *GitRepoSyncer) FetchCommand(ctx context.Context, url string) (cmd *exec
...
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				time.Sleep(100 * time.Millisecond)
+			}
+
+			if strings.Contains(stdout.String(), "Enter password: ") {
```
Comment (ryanslade) [inline cmd/gitserver/server/vcs_syncer.go:174]: nit: Could reverse the check here and reduce indentation

**Review by ryanslade — APPROVED** [review]

File: cmd/gitserver/server/server.go (hunk 5 above)
Comment (unknwon) [reply cmd/gitserver/server/server.go:880]: Yes.. the code before this PR basically discards the modifications of `cmd.Env`, which is important for PerforceDepoSyncer that needs to fill in some env vars for `p4` CLI to work.

File: cmd/gitserver/server/vcs_syncer.go (hunk 6 above)
Comment (tsenart) [reply cmd/gitserver/server/vcs_syncer.go:174]: I think you can set the `P4PASSWD` env variable instead. https://www.perforce.com/manuals/cmdref/Content/CmdRef/P4PASSWD.html

File: cmd/gitserver/server/vcs_syncer.go (hunk 3 above)
Comment (tsenart) [reply cmd/gitserver/server/vcs_syncer.go:198]: Could we not use their ping command? https://www.perforce.com/manuals/cmdref/Content/CmdRef/p4_ping.html#p4_ping

**Review by tsenart — APPROVED** [review]

File: cmd/gitserver/server/vcs_syncer.go (hunk 6 above)
Comment (unknwon) [reply cmd/gitserver/server/vcs_syncer.go:174]: p4 does not accept `P4PASSWD` with plain text password unfortunately, that would make things a lot easier. Despite it's called `PASSWD`, it's in fact _a session-based token of the password_. You can run `p4 login -p` to get it printed and set as `P4PASSWD`, but again it's generated by Perforce Server and _session-based_ (can expire).

The current (default) approach of calling `p4 login` (without `-p`) would make p4 store the token in `~/.p4tickets` so we don't need to take the extra step to store it and p4 uses it automatically.

File: cmd/gitserver/server/vcs_syncer.go (hunk 3 above)
Comment (unknwon) [reply cmd/gitserver/server/vcs_syncer.go:198]: 🤔 I'm not clear what's your intent? Are you saying we should or should use the "ping" command?

"p4 ping" is the cheapest command I found to test if the current session is still valid, and prompts for re-login when not.

Comment (tsenart) [reply cmd/gitserver/server/vcs_syncer.go:198]: Never mind, I misread the command you were using to be login :)

File: cmd/gitserver/server/vcs_syncer.go (hunk 6 above)
Comment (tsenart) [reply cmd/gitserver/server/vcs_syncer.go:174]: So, do we have to start a new session for every command? Seems inefficient. I think we should investigate having the `~/.p4tickets` cache (as a follow up).

Comment (unknwon) [reply cmd/gitserver/server/vcs_syncer.go:174]: >So, do we have to start a new session for every command?

No we don't :D That's why we first try a ping, with minimal cost to test if session-token (stored in `~/.p4tickets` by `p4 login`) is still valid, if so we return directly.

See https://github.com/sourcegraph/sourcegraph/pull/17551/files#diff-3a6fbdbdc8d292ed9a811e2dfbde116dc242e256d4deac0cac2bced73b88a826R193-R196

**Comment by unknwon** [comment]
Thanks everyone, merging 🎉

**Comment by tsenart** [comment]
@unknwon: Is the `P4PASS` env var different than `P4PASSWD`? https://git-scm.com/docs/git-p4#Documentation/git-p4.txt-git-p4password

**Comment by unknwon** [comment]
> @unknwon: Is the `P4PASS` env var different than `P4PASSWD`? [git-scm.com/docs/git-p4#Documentation/git-p4.txt-git-p4password](https://git-scm.com/docs/git-p4#Documentation/git-p4.txt-git-p4password)

Good find! But it has the same restriction that it must be a "ticket value" as pointed out by `p4 help login` for `-P <password>` option:

```
	When a user logs in to Perforce, they are prompted for a password
	If they enter the correct password, they are issued a ticket.  The
	ticket expires when the default timeout value has been reached and
	is valid only for the host machine where the 'login' command was
	executed (see below for exception).

	The ticket can be used anywhere that a password can be used.

	Example: p4 -P <ticket value> changes -m1
```

Admin can configure the ticket timeout, and [the default timeout seems to be 12 hours](https://stackoverflow.com/questions/28758099/p4-password-unset-or-invalid-in-jenkins).
